src/ML/src/velMANN.cpp

```diff
@@ -0,0 +1,354 @@
+                                          m_pimpl->structuredOutput.rawData.data(),
+                                          m_pimpl->structuredOutput.rawData.size(),
+                                          m_pimpl->structuredOutput.shape.data(),
+                                          m_pimpl->structuredOutput.shape.size());
+    // populate variable handler related to the output
+    // the serialization matters
+    m_pimpl->structuredOutput.handler.addVariable("future_base_linear_velocities",
+                                                  3 * (1 + projectedBaseDatapoints / 2));
+    m_pimpl->structuredOutput.handler.addVariable("future_base_angular_velocities",
+                                                  3 * (1 + projectedBaseDatapoints / 2));
+    m_pimpl->structuredOutput.handler.addVariable("joint_positions", numberOfJoints);
+    m_pimpl->structuredOutput.handler.addVariable("joint_velocities", numberOfJoints);
+    m_pimpl->structuredOutput.handler.addVariable("base_position", 3);
+    m_pimpl->structuredOutput.handler.addVariable("base_angle", 3);
+
+    // resize the output
+    m_pimpl->output.futureBaseLinearVelocityTrajectory.resize(3, (1 + projectedBaseDatapoints / 2));
+    m_pimpl->output.futureBaseAngularVelocityTrajectory.resize(3, (1 + projectedBaseDatapoints / 2));
+    m_pimpl->output.jointPositions.resize(numberOfJoints);
+    m_pimpl->output.jointVelocities.resize(numberOfJoints);
+    m_pimpl->output.basePosition.resize(3);
+    m_pimpl->output.baseAngle.resize(3);
+
+    m_pimpl->state = Impl::FSM::Initialized;
+
+    return true;
+}
+
+bool velMANN::setInput(const velMANNInput& input)
+{
+    if (m_pimpl->state != Impl::FSM::Initialized && m_pimpl->state != Impl::FSM::Running)
+    {
+        log()->error("[velMANN::setInput] The network is not initialized, please call initialize()");
+        return false;
+    }
+    return m_pimpl->populateInput(input);
+}
+
+bool velMANN::advance()
+{
+    auto unpackMatrix = [this](const std::string& variableName,
+                               Eigen::Ref<Eigen::MatrixXd> matrix) {
+        const auto& variable = m_pimpl->structuredOutput.handler.getVariable(variableName);
+        assert(variable.isValid());
+
+        // the matrix has been already allocated
+        const std::size_t rows = matrix.rows();
+        const std::size_t cols = matrix.cols();
+
+        matrix = Eigen::Map<const Eigen::MatrixXf>(m_pimpl->structuredOutput.rawData
+                                                       .segment(variable.offset, variable.size)
+                                                       .data(),
+                                                   rows,
+                                                   cols)
+                     .cast<double>();
+
+        return;
+    };
+
+    const char* inputNames[] = {"input"};
+    const char* outputNames[] = {"output"};
+
+    m_pimpl->session->Run(Ort::RunOptions(),
```

Comment:
We may add the possibility of limiting the number of processes used by onnxruntime as done in MANN

Reply:
This has been implemented in 121053069.